File: server/src/HandService.ts
import {
  HandHelper,
  Hand,
  State,
  UnknownCard,
  Player,
  Result,
  Card,
} from './store';
import { CardService, theCardService } from './CardService';
import { UserService, theUserService } from './UserService';

export const DEALER_ID = 'dealer';
export const DEALER = { id: DEALER_ID } as Player;

const helper = new HandHelper();
export class HandService {
  cardService: CardService;
  userService: UserService;

  constructor() {
    this.cardService = theCardService;
    this.userService = theUserService;
  }

  startTurn(state: State): State {
    const playersHands = [];
    var dealersHand = this.create(DEALER_ID, 0);

    dealersHand = this.addCard(dealersHand, UnknownCard);
    const { currentGame } = state;
    for (let playerId of currentGame.players) {
      var player = this.lookupPlayer(state, playerId);
      var hand = this.create(playerId, player.currentBet);
      hand = this.addCard(hand, this.cardService.nextCard());
      hand = this.addCard(hand, this.cardService.nextCard());
      playersHands.push(hand);
    }
    dealersHand = this.addCard(dealersHand, this.cardService.nextCard());
    return this.setActiveHand(
      { ...state, currentGame, playersHands, dealersHand, turnIsGoing: true },
      0
    );
  }

  create(playerId: string, currentBet: number): Hand {
    return {
      player: playerId,
      bet: currentBet,
      cards: [],
      isStaying: false,
      result: Result.PLAYING,
    };
  }

  getActiveHand(state: State): Hand {
    return state.playersHands[state.activeHand];
  }

  hitActiveHand(state: State): State {
    const hand = this.addCard(
      this.getActiveHand(state),
      this.cardService.nextCard()
    );
    var newState = {
      ...state,
      playersHands: this.replace(state.playersHands, state.activeHand, hand),
    };
    if (helper.isDone(hand)) {
      newState = this.nextActiveHand(newState);
    }
    return newState;
  }

  stayActiveHand(state: State): State {
    const hand = this.stay(this.getActiveHand(state));
    var newState = {
      ...state,
      playersHands: this.replace(state.playersHands, state.activeHand, hand),
    };
    return this.nextActiveHand(newState);
  }

  doubleDownActiveHand(state: State): State {
    const hand = this.getActiveHand(state);
    var newState = state;
    if (hand.cards.length === 2) {
      // double down is to give the active hand a card, double the bet, and stay
      var newHand = this.addCard(hand, this.cardService.nextCard());
      newHand = { ...newHand, bet: newHand.bet * 2 };
      newState = {
        ...newState,
        playersHands: this.replace(
          newState.playersHands,
          newState.activeHand,
          newHand
        ),
      };
      newState = this.stayActiveHand(newState);
    }

    return newState;
  }

  nextActiveHand(state: State): State {
    return this.setActiveHand(state, state.activeHand + 1);
  }

  setActiveHand(state: State, activeHand: number) {
    var newState = { ...state, activeHand };
    if (!this.anotherHandExists(newState)) {
      newState = this.giveDealerControl(newState);
    } else {
      var hand = this.getActiveHand(newState);
      if (helper.isDone(hand)) {
        newState = this.nextActiveHand(newState);
      }
    }
    return newState;
  }

  anotherHandExists(state: State): boolean {
    return state.activeHand < state.playersHands.length;
  }

  giveDealerControl(state: State): State {
    console.log('Dealer took control');
    var cards = state.dealersHand.cards.filter((card) => card !== UnknownCard);
    cards.push(this.cardService.nextCard());
    var dealersHand = { ...state.dealersHand, cards };
    return { ...state, dealersHand };
  }

  dealerTakesCard(state: State): State {
    var { dealersHand } = state;
    if (helper.highestTotal(dealersHand) < 17) {
      dealersHand = this.addCard(dealersHand, this.cardService.nextCard());
    }
    return { ...state, dealersHand };
  }

  splitActiveHand(state: State) {
    var newState = state;
    var hand = this.getActiveHand(state);
    if (helper.canSplit(hand)) {
      var hand1 = {
        ...hand,
        cards: [hand.cards[0], this.cardService.nextCard()],
      } as Hand;
      var hand2 = {
        ...hand,
        cards: [hand.cards[1], this.cardService.nextCard()],
      } as Hand;

      var playersHands = newState.playersHands.slice();
      playersHands.splice(newState.activeHand, 1, hand1, hand2);
      newState = this.setActiveHand(
        { ...newState, playersHands },
        newState.activeHand
      );
    }
    return newState;
  }

  endTurn(state: State): State {
    var playersHands = [];
    for (let hand of state.playersHands) {
      hand = this.setResult(hand, this.handResult(hand, state.dealersHand));
      playersHands.push(hand);
      var player = this.lookupPlayer(state, hand.player);
      switch (hand.result) {
        case Result.LOSS:
          state = this.userService.modifyPlayer(
            state,
            this.removeChips(player, hand.bet)
          );
          break;
        case Result.PUSH:
        default:
          break;
        case Result.WIN:
          state = this.userService.modifyPlayer(
            state,
            this.addChips(player, hand.bet)
          );
          break;
        case Result.BLACKJACK:
          state = this.userService.modifyPlayer(
            state,
            this.addChips(player, hand.bet * 1.5)
          );
          break;
      }
    }
    return { ...state, playersHands, turnIsGoing: false };
  }

  handResult(hand: Hand, dealersHand: Hand): Result {
    if (helper.isBusted(hand)) {
      return Result.LOSS;
    } else if (helper.isBlackjack(dealersHand)) {
      return helper.isBlackjack(hand) ? Result.PUSH : Result.LOSS;
    } else if (helper.isBlackjack(hand)) {
      return Result.BLACKJACK;
    } else if (helper.isBusted(dealersHand)) {
      return Result.WIN;
    }
    const dealerTotal = helper.bestTotal(dealersHand);
    const handTotal = helper.bestTotal(hand);
    if (dealerTotal && dealerTotal > handTotal) {
      return Result.LOSS;
    } else if (dealerTotal === handTotal) {
      return Result.PUSH;
    }
    return Result.WIN;
  }

  static create(player: Player, bet: number): Hand {
    return {
      cards: [],
      player: player.id,
      bet,
      isStaying: false,
      result: Result.PLAYING,
    };
  }

  addCard(hand: Hand, card: Card): Hand {
    return { ...hand, cards: [...hand.cards, card] };
  }

  setResult(hand: Hand, result: Result): Hand {
    return { ...hand, result };
  }

  stay(hand: Hand): Hand {
    return { ...hand, isStaying: true };
  }

  replace<T>(arr: Array<T>, index: number, replaceItem: T) {
    return arr.map((item, i) => {
      if (i !== index) {
        // This isn't the item we care about - keep it as-is
        return item;
      }

      return replaceItem;
    });
  }

  public addChips(player: Player, chips: number): Player {
    var chipBalance = player.chipBalance + chips;
    console.log(
      'Adding ' + chips + ' to ' + player.name + ' new balance=' + chipBalance
    );
    return { ...player, chipBalance };
  }

  public removeChips(player: Player, chips: number): Player {
    if (player.chipBalance >= chips) {
      return this.addChips(player, -chips);
    }
    return player;
  }

  lookupPlayer(state: State, id: string): Player {
    var player = state.activeUsers.find((player) => player.id === id);
    if (player === undefined) {
      throw 'Something bad happened, could not find player ' + id;
    }
    return player;
  }
}

export const theHandService = new HandService();

File: django/blackjack/store/hand_service.py
import math
from blackjack.state.card import UNKNOWN_CARD, Card
from blackjack.state.hand import Hand, HandHelper, Result
from blackjack.state.player import Player
from blackjack.state.state import State
from blackjack.store.card_service import CardService
from blackjack.store.user_service import UserService


DEALER_ID = 'dealer'

class HandService:
    def __init__(self, cardService: CardService, userService: UserService) -> None:
        self.cardService = cardService
        self.userService = userService
        self.helper = HandHelper()

    def startTurn(self, state: State) -> State: 
        playersHands: list[Hand] = []
        dealersHand = self.create(DEALER_ID, 0)
        dealersHand = self.addCard(dealersHand, UNKNOWN_CARD)
        currentGame = state.currentGame

        for player in currentGame.players:
            hand = self.create(player.id, player.currentBet)
            hand = self.addCard(hand, self.cardService.nextCard())
            hand = self.addCard(hand, self.cardService.nextCard())
            playersHands.append(hand)

        dealersHand = self.addCard(dealersHand, self.cardService.nextCard())

        newState = state.clone()
        newState.currentGame = currentGame
        newState.playersHands = playersHands
        newState.dealersHand = dealersHand
        newState.turnIsGoing = True
        newState.activeUsers = state.activeUsers
        return self.setActiveHand(newState, 0)

    def create(self, playerId: str, currentBet: int) -> Hand :
        newHand = Hand()
        newHand.player = playerId
        newHand.bet = currentBet
        return newHand

    def getActiveHand(self, state: State) -> Hand: 
        return state.playersHands[state.activeHand]

    def hitActiveHand(self, state: State) -> State: 
        hand = self.addCard(self.getActiveHand(state), self.cardService.nextCard())
        newState = state.clone()
        newState.playersHands = self.replace(state.playersHands, state.activeHand, hand)
        if self.helper.isDone(hand):
            newState = self.nextActiveHand(newState)
        return newState

    def stayActiveHand(self, state: State) -> State: 
        hand = self.stay(self.getActiveHand(state))
        newState = state.clone()
        newState.playersHands = self.replace(state.playersHands, state.activeHand, hand)
        return self.nextActiveHand(newState)

    def doubleDownActiveHand(self, state: State) -> State: 
        hand = self.getActiveHand(state)
        newState = state
        if len(hand.cards) == 2:
            # double down is to give the active hand a card, double the bet, and stay
            newHand = self.addCard(hand, self.cardService.nextCard())
            newHand.bet = newHand.bet * 2
            newState = newState.clone()
            newState.playersHands =  self.replace(newState.playersHands, newState.activeHand, newHand) 
            newState = self.stayActiveHand(newState)

        return newState

    def nextActiveHand(self, state: State) -> State: 
        return self.setActiveHand(state, state.activeHand + 1)

    def setActiveHand(self, state: State, activeHand: int) -> State:
        newState = state.clone()
        newState.activeHand = activeHand
        if not self.anotherHandExists(newState):
            newState = self.giveDealerControl(newState)
        else:
            hand = self.getActiveHand(newState)
            if self.helper.isDone(hand):
                newState = self.nextActiveHand(newState)

        return newState

    def anotherHandExists(self, state: State) -> bool : 
        return state.activeHand < len(state.playersHands)

    def giveDealerControl(self, state: State) -> State : 
        print('Dealer took control')
        
        # remove the unknown card and add a real one
        cards = list(filter(lambda card: card != UNKNOWN_CARD, state.dealersHand.cards))
        cards.append(self.cardService.nextCard())

        dealersHand = state.dealersHand.clone()
        dealersHand.cards = cards
        
        newState = state.clone()
        newState.dealersHand = dealersHand
        if self.helper.isBlackjack(dealersHand):
            newState = self.endTurn(newState)

        return newState

    def dealerTakesCard(self, state: State) -> State :
        dealersHand = state.dealersHand
        if self.helper.highestTotal(dealersHand) < 17:
            dealersHand = self.addCard(dealersHand, self.cardService.nextCard())

        newState = state.clone()
        newState.dealersHand = dealersHand
        return newState

    def splitActiveHand(self, state: State) -> State : 
        newState = state
        hand = self.getActiveHand(state)
        if self.helper.canSplit(hand):
            hand1 = hand.clone()
            hand1.cards = [hand.cards[0], self.cardService.nextCard()]

            hand2 = hand.clone()
            hand2.cards = [hand.cards[1], self.cardService.nextCard()]

            newState = state.clone()
            newState.playersHands = newState.playersHands[0:newState.activeHand] + [hand1,hand2] + newState.playersHands[newState.activeHand:]
        return newState

    def endTurn(self, state: State) -> State:
        print('End turn')
        playersHands: list[Hand] = []
        for hand in state.playersHands: 
            hand = self.setResult(hand, self.handResult(hand, state.dealersHand))
            playersHands.append(hand)
            player = self.lookupPlayer(state, hand.player)
            match hand.result:
                case Result.Loss:
                    state = self.userService.modifyPlayer(state, self.removeChips(player, hand.bet))
                case Result.Win:
                    state = self.userService.modifyPlayer(state, self.addChips(player, hand.bet))
                case Result.Blackjack:
                    state = self.userService.modifyPlayer(state, self.addChips(player, math.trunc(hand.bet * 1.5)))
                case Result.Push:
                    pass
                case Result.Playing:
                    pass

        newState = state.clone()
        newState.turnIsGoing = False
        return newState

    def handResult(self, hand: Hand, dealersHand: Hand) -> Result:  
        if self.helper.isBusted(hand):
            return Result.Loss
        elif self.helper.isBlackjack(dealersHand): 
            return Result.Push if self.helper.isBlackjack(hand) else Result.Loss
        elif self.helper.isBlackjack(hand):
            return Result.Blackjack
        elif self.helper.isBusted(dealersHand): 
            return Result.Win
        
        dealerTotal = self.helper.bestTotal(dealersHand)
        handTotal = self.helper.bestTotal(hand)
        if dealerTotal and dealerTotal > handTotal:
            return Result.Loss
        elif dealerTotal == handTotal: 
            return Result.Push
        return Result.Win

    def addCard(self, hand: Hand, card: Card) -> Hand :
        newHand = hand.clone()
        newHand.cards = hand.cards.copy()
        newHand.cards.append(card)
        return newHand

    def setResult(self, hand: Hand, result: Result) -> Hand :
        newHand = hand.clone()
        newHand.result = result
        return newHand

    def stay(self, hand: Hand) -> Hand :
        newHand = hand.clone()
        newHand.isStaying = True
        return newHand

    def replace(self, arr: list[Hand], index: int, replaceItem: Hand) -> list[Hand]:
        newArray: list[Hand] = []
        for i in range(0, len(arr)):
            newArray.append(replaceItem if i == index else arr[i])
        return newArray

    def addChips(self, player: Player, chips: int) -> Player :
        chipBalance = player.chipBalance + chips
        print(f'Adding {chips} to {player.name} new balance={chipBalance}')

        newPlayer = player.clone()
        newPlayer.chipBalance = chipBalance
        return newPlayer

    def removeChips(self, player: Player, chips: int) -> Player:
        if player.chipBalance >= chips:
            return self.addChips(player, -chips)
        return player

    def lookupPlayer(self, state: State, id: str) -> Player: 
        player = next(filter(lambda p: p.id == id, state.activeUsers), None)
        if player == None:
            raise Exception('Something bad happened, could not find player ' + id)
        return player